Mathlib/Analysis/CStarAlgebra/CStarVec.lean

```diff
@@ -0,0 +1,329 @@
+/-
+Copyright (c) 2024 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import Mathlib.Analysis.CStarAlgebra.Module
+
+/-!
+# Vectors with entries in a C⋆-algebra
+
+This file creates a type copy of `n → A` meant for vectors with entries in a C⋆-algebra `A`. It is
+endowed with a Hilbert C⋆-module structure, which in particular comes with an `A`-valued
+"inner product" `⟪w, v⟫ = ∑ i, star (w i) * v i` and a norm `‖v‖ = √‖⟪v, v⟫‖`.
+
+## Main declarations
+
++ `CStarVec n A`: the type copy
++ `instCStarModule`: the Hilbert C⋆-module instance
+
+## Implementation notes
+
+The norm on this type induces the product uniformity and bornology, but these are not defeq to
+`Pi.uniformSpace` and `Pi.instBornology`. Hence, we prove the equality to the Pi instances and
+replace the uniformity and bornology by the Pi ones when registering the
+`NormedAddCommGroup (CStarVec n A)` instance. See the docstring of the `TopologyAux` section below
+for more details.
+
+-/
+
+open scoped ComplexOrder RightActions Topology Uniformity Bornology
+
+/-- A vector with entries in a C⋆-algebra and a Hilbert C⋆-module structure. -/
+def CStarVec (n : Type*) (A : Type*) := n → A
```

Comment:
Yeah, I'm still not too sure what the best approach is here. I went with `irreducible` for `E →WOT[𝕜] F`, but then I saw that e.g. `PiLp` is even an abbrev to inherit all the instances from `WithLp`, and it doesn't seem to cause major problems (but maybe I just didn't see them). I actually went with `abbrev` for `CStarMatrix` also to inherit everything from matrices, basically copying the `PiLp` design.

Reply:
Well, the `PiLp` bit is basically just to have nicer syntax. It's `WithLp` that is the type synonym barrier. That is a default def, neither abbrev nor irreducible. This is, essentially exactly what caused the issues in `EuclideanSpace` which is itself another `abbrev` for a `WithLp` variant. The problem is that Lean would see through `EuclideanSpace K n`, which is abbreviated for `WithLp 2 (n -> K)`, to bare functions `n -> K` if the user was ever so slightly sloppy. My thinking is that this is basically the same situation. So I'm convincing myself to advocate for irreducible. Especially success we can avoid pain whenever we need to with `unseal`. This way we keep the barrier tight and force users to use `ofFun` and the `CoeFun` instance, which is what they should be doing already to avoid defeq abuse.

Reply:
I tried to convert this to `irreducible` and it's surprisingly painful. You can't just unseal every lemma that gives you trouble (stuff like `zero_apply` and friends), because once those things are defined via defeq with functions, you can't use them later when resealed. So you have to redo the whole hierarchy of `SubNegMonoid` and friends by hand with `ofFun`. I think this would work fine if we could state lemmas in "sealed mode" and then unseal only in the proof.

Reply:
Aha, I see. Thanks for trying.